1_prep.R

```diff
@@ -157,6 +157,7 @@ p1 <- list(
     p1_obs_buffer_from_dam_rds,
     subset_model_obs_data(data = p1_obs_buff_sf,
                           site_id = p1_nldas_site_ids,
+                          remove_dups = TRUE,
```

Comment:
Also commenting about [this target](https://github.com/USGS-R/lake-temperature-missouri-models/pull/14/files#diff-439fd71d5440a058f307cd26a77a90f839174d1c585d7796c319e7f88c8a77d8L114) because it is no longer tracking file changes when you put the file names into a tibble. That is why we'd either continue the "files" tag for targets, or pair the file entries in the tibble with their respective hashes when you create that tibble target, e.g.,
```r
tar_target(p1_obs_coop_files,
             tibble(
                 temp_files = p1_obs_coop_tmp_data,
                 xwalk_files = p1_obs_coop_xwalks_rds[c(1, 2, 4, 3, 6, 5)]
               ) %>% mutate(temp_file_hash = tools::md5sum(temp_files), xwal_files = tools::md5sum(xwalk_files))
             )
```

Viznetwork looks up to date (ignore _buff and beyond, as I haven't built those yet in my review):
<img width="727" alt="image" src="https://user-images.githubusercontent.com/2349007/186684869-2b17eb3e-c67a-482c-9c66-3922c8aecea7.png">

But, w/o tracking file changes, if I change one of the files (which is possible, since it is in `in/`), I see the expected targets downstream are out of date now (good!): 
<img width="746" alt="image" src="https://user-images.githubusercontent.com/2349007/186685004-a46cbc2e-d31b-48d5-84ea-213f5cab8ab7.png">

But if I selectively build that tibble I mentioned above with `tar_make('p1_obs_coop_files')`, targets now thinks the downstream targets are still fresh even though the observational data has changed:
<img width="758" alt="image" src="https://user-images.githubusercontent.com/2349007/186685130-f6a54785-9d42-4c50-8cd7-f3cdb1a69668.png">

This is because targets is now tracking the _tibble_ for changes, not the files themselves. So since the tibble is the same files in the same order, it looks unchanged.

Reply:
This comment is a follow up to a conversation we had earlier this morning - 

Two things:
 * re: brittleness - I hear what you're saying about the brittleness of `p1_obs_coop_files` - I'll swap it out for hard coding. That being said - I
 * re: tracking - Ahhhhh, ok, this is definitely tying back into our conversation from last Friday on a different part of the pipeline. The key lesson here is that if you are building a tibble to iterate over you need to include hashes or it will lose the tracking.

I will make these two updates: hard coding the relationships between `temp_files` and `xwalk_files` and adding the hash to this target.

Reply:
I am running into a problem when I update this target as we discussed. Here is an outline of the problem - 

## Original target:
```r
tar_target(p1_obs_coop_files,
             tibble(
                 temp_files = p1_obs_coop_tmp_data,
                 xwalk_files = p1_obs_coop_xwalks_rds[c(1, 2, 4, 3, 6, 5)]
               ) %>% mutate(temp_file_hash = tools::md5sum(temp_files), xwal_files = tools::md5sum(xwalk_files))
             )
```

![image](https://user-images.githubusercontent.com/15007294/186739023-f2201de1-9920-4f4b-8f28-f4ec83edcb5a.png)

and replacing lat/longs with zero in Bull_Shoals_and_LOZ_profile_data_LMVP_latlong.csv:
![image](https://user-images.githubusercontent.com/15007294/186739065-ec4128f6-9a03-4996-b73b-b7d1b2acf637.png)



## Updated target

```r
tar_target(p1_obs_coop_files,
             tibble(
               temp_files = c(
                 '1_prep/in/obs_review/temp/Bull_Shoals_and_LOZ_profile_data_LMVP.rds',
                 '1_prep/in/obs_review/temp/Bull_Shoals_Lake_DO_and_Temp.rds',
                 '1_prep/in/obs_review/temp/Missouri_USACE_2009_2021.rds',
                 '1_prep/in/obs_review/temp/Temp_DO_BSL_MM_DD_YYYY.rds',
                 '1_prep/in/obs_review/temp/UniversityofMissouri_LimnoProfiles_2017_2020.rds',
                 '1_prep/in/obs_review/temp/Waterbody_Temperatures_by_State.rds'
                 ),
               xwalk_files = c(
                 '1_prep/in/obs_review/spatial/Bull_Shoals_and_LOZ_profile_data_LMVP_latlong_sf.rds',
                 '1_prep/in/obs_review/spatial/Bull_Shoals_Lake_DO_and_Temp_latlong_sf.rds',
                 '1_prep/in/obs_review/spatial/mo_usace_sampling_locations_sf.rds',
                 '1_prep/in/obs_review/spatial/Temp_DO_BSL_MM_DD_YYYY_latlong_sf.rds',
                 '1_prep/in/obs_review/spatial/UniversityofMissouri_2017_2020_Profiles_sf.rds',
                 '1_prep/in/obs_review/spatial/Navico_lakes_depths_sf.rds'
                 )
               ) %>% mutate(temp_file_hash = tools::md5sum(temp_files),
                            xwalk_file_hash = tools::md5sum(xwalk_files)
                            )
             )
```

![image](https://user-images.githubusercontent.com/15007294/186739172-44bb5711-96eb-41e8-ad3f-a7a900b2b668.png)

![image](https://user-images.githubusercontent.com/15007294/186739183-f7c99833-54d5-46ae-980d-0cdb633cc337.png)

If I update one of the csvs in `p1_obs_manual_xwalks_csv` (e.g., replacing lat/longs with zero in Bull_Shoals_and_LOZ_profile_data_LMVP_latlong.csv), `p1_obs_coop_files` will not update.

More later, I have to hop to another task.

Reply:
I'm sorry, I could've been better about explaining this potential snag. 

If hard-coding the file paths, you can still take advantage of targets' tracking of the "file" type, by having two file vectors as targets that have "file" type. Any changes to those raw files would be picked up by targets. 

The downstream tibble target to combine them could be

`tibble(temp_files = temp_file_vector, xwalk_files = xwalk_file_vector)` then the mutate to attach hashes to this new tibble target. 

Can explain in more detail why this is necessary.

Reply:
Ok more now:

This is needed because hard-coding the file paths _and_ calculating the hashes in the same target will only happen once, which is during the intial build. In the future if the file contents change, there will be nothing that tells targets that the tibble needs to be rebuilt (and the mutate called again to set a new hash value). So in that case, the change in file doesn't get passed along downstream. 

You can solve this problem by using two file vectors with the "file" type as mentioned above, then another target that combines them and runs the mutate. This works because any change in file contents to either vector will trigger a rebuild of the tibble and a rebuild of that tibble will update the mutate and the hash value. 

However, you can make this work with your updated tibble that does this in a smaller number of targets (1 instead of 3) but you'd need to add `cue = 'always'` to your `tar_target()` params. This causes the target to get rebuilt everytime. In this case, that is safe and also inexpensive, so might be worth considering. One reason _not_ to use that pattern is that we're seeing the files-as-vector/data.frame issue pop up a lot with the consequence of losing file contents tracking, so if someone copied your pattern but left out the `cue`, it would be brittle. I'm good with either approach though.

Reply:
Thanks @lindsayplatt for the `cue` reminder!

Reply:
Here is the [edit](https://github.com/padilla410/lake-temperature-missouri-models/blob/f15ca26a9990935f284b2b0d780004948cd516ab/1_prep.R#L119-L128)

Here is what I was able to do:
* I did not modify `p1_obs_coop_xwalks_rds` because, if I do so, it will lose the connection to `p1_obs_manual_xwalks_csv` which tracks the manual crosswalks I had to create (see [this chunk](https://github.com/padilla410/lake-temperature-missouri-models/blob/f15ca26a9990935f284b2b0d780004948cd516ab/1_prep.R#L81-L104))
* I hard-coded the file names in [`p1_obs_coop_tmp_rds`](https://github.com/padilla410/lake-temperature-missouri-models/blob/f15ca26a9990935f284b2b0d780004948cd516ab/1_prep.R#L107-L117) (I also renamed this target according to the std convention). Because I cannot hard-code `p1_obs_coop_xwalks_rds`, I hard-coded  `p1_obs_coop_tmp_rds` so that it will properly match `p1_obs_coop_xwalks_rds` when combined into a tibble in `p1_obs_coop_files`

Reply:
I'm good with this moving forward but in the future the "in" folder needs to only contain manually added raw data and any file that is created as a result of an automated target build is in "out" (or other relevant location, just not "in"). The mix of automated and manually added files in `1_prep/in/obs_review/spatial` is challenging to separate and also makes for a more complex target in `p1_obs_coop_xwalks_rds` that you need to rely on being unchanged so that the hard-coded `p1_obs_coop_tmp_rds` file order still aligns with the output from the xwalks_rds target. 

I think you rely on 
```
p1_obs_coop_xwalks_rds
[1] "1_prep/in/obs_review/spatial/Bull_Shoals_and_LOZ_profile_data_LMVP_latlong_sf.rds"
[2] "1_prep/in/obs_review/spatial/Bull_Shoals_Lake_DO_and_Temp_latlong_sf.rds"         
[3] "1_prep/in/obs_review/spatial/Temp_DO_BSL_MM_DD_YYYY_latlong_sf.rds"               
[4] "1_prep/in/obs_review/spatial/mo_usace_sampling_locations_sf.rds"                  
[5] "1_prep/in/obs_review/spatial/Navico_lakes_depths_sf.rds"                          
[6] "1_prep/in/obs_review/spatial/UniversityofMissouri_2017_2020_Profiles_sf.rds"  
```
being fixed, but [1], [2], and [3] from ☝️ come from transforming their respective csvs in `1_prep/in/obs_review/spatial/csvs` to .rds (i.e., they are automatically created with `create_missing_xwalk()`) and [4], [5], and [6] were manually added to the `1_prep/in/obs_review/spatial` folder. 

I recognize this is a tricky situation because you need to effectively do a manual join between the xwalks and the temperature data, which is what you are doing when expecting the two vectors (`p1_obs_coop_tmp_rds` and `p1_obs_coop_xwalks_rds`) to be the same length and in the right order (although including them both in the `p1_obs_coop_files` tibble will cause a useful error if they are ever different lengths in the future). 

In situations like this where you are relying on the result of a `list.files()` to not change, I'd suggest putting in a small check to verify the right files are getting joined. It just feels too risky to have a join be based on order vs a key or some other more direct/explicit method.